### PR TITLE
Add 'Review assessment summaries' feature

### PR DESCRIPTION
**Pull request by allbecauseyoutoldmeso**
### Description of change

Add 'Review assessment summaries' feature.

Sorry, this PR is huge!

The main changes are:

1. Updating the `status` of `assessment_details` to have `assessment_in_progress`, `assessment_complete`, `review_in_progress` and`review_complete`.
2. Adding a `review_status` column to `assessment_details` which reflects the reviewer's verdict.
3. The`AssessmentDetailsReview` form object, which handles all the logic around rendering and submitting the form for this feature (see screenshot).

### Screenshot
Screenshot 2022-11-24 at 16 17 53

### Story Link

https://trello.com/c/GFDrRaCK/1254-reviewer-is-able-to-mark-items-that-need-reviewing-from-assessments-and-can-comment-on-individual-things-that-need-amending

### Decisions 

I experimented with nested forms, but in the end decided that of two complex options a form object was marginally less complex, at at least keeps the complexity in a model and not in the view and controller 😬

### CODE REVIEW

File: app/components/status_tags/reviewer_assessment_detail_component.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    def initialize(assessment_detail:)
+      @assessment_detail = assessment_detail
+    end
+
+    private
+
+    attr_reader :assessment_detail
+
+    def status
+      :updated if assessment_detail.updated?
```
Comment (allbecauseyoutoldmeso) [inline app/components/status_tags/reviewer_assessment_detail_component.rb:14]: There's another ticket which the design team are working on to clarify the other statuses that should show up here:

https://trello.com/c/DdrC2wdW/1344-review-assessment-summary-page-status-tags

File: app/models/planning_application.rb [hunk 2]
```diff
@@ -458,9 +458,23 @@ def planning_history_enabled?
...
+      category: category,
+      status: :review_complete,
+      review_status: :rejected
+    ).last
+  end
+
   AssessmentDetail.categories.each_key do |category|
     define_method(category) do
-      assessment_details.find_by(category: category)
+      assessment_details.where(category: category).first
```
Comment (allbecauseyoutoldmeso) [inline app/models/planning_application.rb:473]: Ensure this returns the most recent record (they're ordered by `created_at_desc`)

File: app/models/assessment_details_review.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      :entry,
+      "entry=",
+      to: assessment_detail,
+      prefix: true
+    )
+
+    delegate(:text, "text=", to: "#{assessment_detail}_comment", prefix: true)
+
+    validates(
+      "#{assessment_detail}_review_status",
```
Comment (notapatch) [inline app/models/assessment_details_review.rb:49]: OK I don't know what this does.
*I'm in danger meme*

File: app/models/assessment_details_review.rb [hunk 4]
```diff
@@ -0,0 +1,131 @@
+# frozen_string_literal: true
+
```
Comment (notapatch) [inline app/models/assessment_details_review.rb:2]: A form object! I will take kate on her word as I don't have the skill to review this

**Review by notapatch — APPROVED** [review]
Approved... comments were through slack as it was long PR ... it sped up PR process

File: app/models/assessment_details_review.rb (hunk 4 above)
Comment (allbecauseyoutoldmeso) [reply app/models/assessment_details_review.rb:2]: I thought I'd leave a bit of an explanation here for posterity!

This is a form object. See [this blog post](https://thoughtbot.com/blog/activemodel-form-objects) for a good introduction to form objects.

The basic idea is that by `ActiveModel::Model` we can make a plain Ruby object active like a Rails Active Model.

File: app/models/assessment_details_review.rb [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  before_save :set_statuses
+
+  ASSESSMENT_DETAILS.each do |assessment_detail|
+    define_method(assessment_detail) do
+      memoize(
+        assessment_detail,
+        planning_application.send("existing_or_new_#{assessment_detail}")
+      )
+    end
```
Comment (allbecauseyoutoldmeso) [inline app/models/assessment_details_review.rb:28]: This parsed to eg
```
def summary_of_works
  @summary_of_works ||= planning_application.summary_of_works
end
```

File: app/models/assessment_details_review.rb [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    define_method("#{assessment_detail}_comment") do
+      memoize(
+        "#{assessment_detail}_comment",
+        send(assessment_detail).existing_or_new_comment
+      )
+    end
```
Comment (allbecauseyoutoldmeso) [inline app/models/assessment_details_review.rb:35]:
```
def summary_of_works_comment
  @summary_of_works_comment ||= summary_of_works.existing_or_new_comment
end
```

File: app/models/assessment_details_review.rb (hunk 3 above)
Comment (allbecauseyoutoldmeso) [reply app/models/assessment_details_review.rb:49]:
```
validates :summary_of_works_review_status, presence: true, if: :status_complete?`
```

File: app/models/assessment_details_review.rb [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    validates(
+      "#{assessment_detail}_entry",
+      presence: true,
+      if: "#{assessment_detail}_review_status_edited_and_accepted?".to_sym
+    )
+
+    validate(
+      "#{assessment_detail}_entry_changed".to_sym,
+      if: "#{assessment_detail}_review_status_edited_and_accepted?".to_sym
+    )
```
Comment (allbecauseyoutoldmeso) [inline app/models/assessment_details_review.rb:69]: `validate :summary_of_work_entry_changed`, if: `summary_of_work_review_status_edited_and_accepted?`

File: app/models/assessment_details_review.rb [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    define_method("#{assessment_detail}_review_status_edited_and_accepted?") do
+      send("#{assessment_detail}_review_status") == "edited_and_accepted"
+    end
+
+    define_method("#{assessment_detail}_entry_changed") do
+      return if !send(assessment_detail).review_status_changed? || send(assessment_detail).entry_changed?
+
+      errors.add("#{assessment_detail}_entry", :not_changed)
+    end
+  end
```
Comment (allbecauseyoutoldmeso) [inline app/models/assessment_details_review.rb:84]:
```
def summary_of_work_entry_changed
  return if !summary_of_work.review_status_changed? || summary_of_work.entry_changed?

  errors.add(:summary_of_work_entry, :not_changed)
end
```

File: app/models/assessment_details_review.rb [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    end
+
+    delegate(
+      :review_status,
+      "review_status=",
+      :entry,
+      "entry=",
+      to: assessment_detail,
+      prefix: true
+    )
```
Comment (allbecauseyoutoldmeso) [inline app/models/assessment_details_review.rb:44]: Creates eg `summary_of_work_review_status` and `summary_of_work_entry` getter and setter methods. 

`summary_of_work_review_status` will return `summary_of_work.review_status` and ``summary_of_work_review_status = x` will set `summary_or_work.review_status` to `x`.

**Pull request merged**